File: SKILLED-SPM/Skilled/Models/Payment.swift
import Foundation
import FirebaseFirestore

enum PaymentStatus: String, Codable {
    case pending = "pending"
    case processing = "processing"
    case completed = "completed"
    case failed = "failed"
    case refunded = "refunded"
    
    var displayText: String {
        switch self {
        case .pending: return "Pending"
        case .processing: return "Processing"
        case .completed: return "Completed"
        case .failed: return "Failed"
        case .refunded: return "Refunded"
        }
    }
}

enum PaymentMethod: String, Codable {
    case creditCard = "creditCard"
    case debitCard = "debitCard"
    case applePay = "applePay"
    case bankTransfer = "bankTransfer"
}

struct SavedPaymentMethod: Codable {
    var type: PaymentMethod
    var last4: String?
    var isDefault: Bool
    var cardBrand: String?
}

struct Payment: Codable {
    let id: String
    let bookingId: String
    let userId: String
    let providerId: String
    let amount: Double
    var status: PaymentStatus
    let paymentMethod: PaymentMethod
    let transactionId: String?
    let createdAt: Date
    var updatedAt: Date
    
    // Convert to dictionary for Firestore
    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "id": id,
            "bookingId": bookingId,
            "userId": userId,
            "providerId": providerId,
            "amount": amount,
            "status": status.rawValue,
            "paymentMethod": paymentMethod.rawValue,
            "createdAt": Timestamp(date: createdAt),
            "updatedAt": Timestamp(date: updatedAt)
        ]
        
        if let transactionId = transactionId {
            dict["transactionId"] = transactionId
        }
        
        return dict
    }
    
    // Create from Firestore document
    static func fromDictionary(_ dict: [String: Any]) -> Payment? {
        guard let id = dict["id"] as? String,
              let bookingId = dict["bookingId"] as? String,
              let userId = dict["userId"] as? String,
              let providerId = dict["providerId"] as? String,
              let amount = dict["amount"] as? Double,
              let statusRaw = dict["status"] as? String,
              let status = PaymentStatus(rawValue: statusRaw),
              let paymentMethodRaw = dict["paymentMethod"] as? String,
              let paymentMethod = PaymentMethod(rawValue: paymentMethodRaw) else {
            return nil
        }
        
        // Handle dates
        let createdAt: Date
        if let timestamp = dict["createdAt"] as? Timestamp {
            createdAt = timestamp.dateValue()
        } else {
            return nil
        }
        
        let updatedAt: Date
        if let timestamp = dict["updatedAt"] as? Timestamp {
            updatedAt = timestamp.dateValue()
        } else {
            return nil
        }
        
        // Optional fields
        let transactionId = dict["transactionId"] as? String
        
        return Payment(
            id: id,
            bookingId: bookingId,
            userId: userId,
            providerId: providerId,
            amount: amount,
            status: status,
            paymentMethod: paymentMethod,
            transactionId: transactionId,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }
}

File: Skilled.Data/Models/Payment.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Skilled.Data.Models;

public enum PaymentStatus
{
    Pending,
    Processing,
    Completed,
    Failed,
    Refunded,
    Disputed
}

public enum PaymentMethod
{
    CreditCard,
    DebitCard,
    BankTransfer,
    PayPal,
    ApplePay,
    GooglePay,
    Cash,
    Other
}

public class Payment
{
    public Guid Id { get; set; }

    public Guid UserId { get; set; }
    public Guid ProviderId { get; set; }

    /// <summary>Link to the booking this payment covers.</summary>
    public Guid? BookingId { get; set; }

    [Column(TypeName = "decimal(18,2)")]
    public decimal Amount { get; set; }

    [Column(TypeName = "decimal(18,2)")]
    public decimal RefundAmount { get; set; } = 0;

    public DateTime Date { get; set; } = DateTime.UtcNow;

    public PaymentStatus Status { get; set; } = PaymentStatus.Pending;

    [Required]
    public PaymentMethod Method { get; set; }

    [MaxLength(100)]
    public string? TransactionId { get; set; }

    [MaxLength(500)]
    public string? Notes { get; set; }

    public DateTime? CompletedAt { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    // ── Navigation properties ────────────────────────────────────────────────
    public virtual User? User { get; set; }
    public virtual ServiceProvider? Provider { get; set; }
    public virtual Booking? Booking { get; set; }
}
